mtt/src/mtt_types.rs

```diff
@@ -0,0 +1,769 @@
+    fn set_4k_page_memory_type(&mut self, memory_type: MttMemoryType) -> Result<()> {
+        match self.current_l2_entry {
+            MttL1Dir(_) => {
+                let (l1_index, sub_index) = Self::get_l1_index_and_subindex(self.current_phys_addr);
+                let mtt_l1_page_pool_len = self.inner.mtt_l1_page_pool_base.len();
+                let l1_pool_index =
+                    Self::get_l1_page_pool_index(self.current_phys_addr, mtt_l1_page_pool_len)?;
+                let mtt_l1_page = &mut self.inner.mtt_l1_page_pool_base[l1_pool_index].base[0..];
+                let value = mtt_l1_page[l1_index];
+                let bit_value = if memory_type == MttMemoryType::NonConfidential {
+                    0u64
+                } else {
+                    1u64
+                };
+                // Clear the original 2-bits and OR in the new value
+                let value = (value & !(3 << sub_index)) | (bit_value << sub_index);
+                mtt_l1_page[l1_index] = value;
+                Ok(())
+            }
+            _ => Err(Error::Non4KL2Entry),
+        }
+    }
+
+    // Returns the memory type for physical address.
+    // The page can be mapped at any granularity in the MTT.
+    fn get_memory_type(&self) -> Result<MttMemoryType> {
+        match self.current_l2_entry {
+            MttNonConfidential1G => Ok(MttMemoryType::NonConfidential),
+            MttConfidential1G => Ok(MttMemoryType::Confidential),
+            Mtt2MBPages(value) => {
+                let addr_bits = self.get_phys_addr().bits();
+                let aligned_2mb = PageSize::Size2M.round_down(addr_bits);
+                let aligned_64mb = addr_bits & !((1024 * 1024 * 64) - 1);
+                let bit_index = (aligned_2mb - aligned_64mb) / PageSize::Size2M as u64;
+                if (value & (1 << bit_index)) == MttMemoryType::Confidential as u64 {
+                    Ok(MttMemoryType::Confidential)
+                } else {
+                    Ok(MttMemoryType::NonConfidential)
+                }
+            }
+            MttL1Dir(_) => self.get_4k_page_memory_type(),
+        }
+    }
+
+    // Performs checks to ensure that the operation on the entire memory range
+    // can succeed.
+    fn validate_page_range(&self) -> Result<()> {
+        let phys_addr = self.current_phys_addr;
+        let page_size = self.page_size;
+        let num_pages = self.num_pages;
+        // TODO: Add support for non-4K pages
+        if page_size.is_huge() {
+            return Err(UnsupportedPageSize);
+        }
+        // Ensure that that range is valid.
+        let mut page_iter = phys_addr
+            .iter_from_with_size(page_size)
+            .ok_or(InvalidAddress)?;
+        let last_page = if num_pages > 0 {
+            page_iter.nth(num_pages - 1)
+        } else {
+            None
+        };
+        if last_page.is_none() {
+            return Err(InvalidAddressRange);
+        }
```

Comment:
Does this whole block from 504-515 reduce to:
```
if num_pages == 0 {
    return Err(InvalidAddressRange);
}
```
?

Reply:
Think a simple check for num_pages == 0 won't address the nth page check for a valid range. For example, the start address can be the top of the physical address range - 4096, and if num_pages is anything > 1, we'll overflow.